main/lsp/requests/document_symbol.cc

```diff
@@ -102,29 +102,68 @@ unique_ptr<ResponseMessage> DocumentSymbolTask::runRequest(LSPTypecheckerInterfa
         {core::SymbolRef::Kind::TypeArgument, gs.typeArgumentsUsed()},
         {core::SymbolRef::Kind::TypeMember, gs.typeMembersUsed()},
     };
+    vector<core::SymbolRef> candidates;
     for (auto [kind, used] : symbolTypes) {
         for (uint32_t idx = 1; idx < used; idx++) {
             core::SymbolRef ref(gs, kind, idx);
             if (hideSymbol(gs, ref)) {
                 continue;
             }
 
-            // a bit counter-intuitive, but this actually should be `!= fref`, as it prevents duplicates.
-            if (ref.owner(gs).loc(gs).file() != fref || ref.owner(gs) == core::Symbols::root()) {
-                for (auto definitionLocation : ref.locs(gs)) {
-                    if (definitionLocation.file() != fref) {
-                        continue;
-                    }
-
-                    auto data = symbolRef2DocumentSymbol(gs, ref, fref);
-                    if (data) {
-                        result.push_back(move(data));
-                        break;
-                    }
+            // If the owner lives in this file, then the owner will be caught elsewhere
+            // in this loop and `ref` will be output as a child of owner.  So we should
+            // avoid outputting `ref` at this point.  Symbols owned by root, however,
+            // should always be output, as we don't output root itself.
+            if (ref.owner(gs).loc(gs).file() == fref && ref.owner(gs) != core::Symbols::root()) {
+                continue;
+            }
+
+            for (auto definitionLocation : ref.locs(gs)) {
+                if (definitionLocation.file() != fref) {
+                    continue;
                 }
+
+                candidates.emplace_back(ref);
             }
         }
     }
+
+    // Despite our best efforts above, we might still output duplicates from the
+    // list of candidate symbols.  For instance, given an ownership chain:
+    //
+    // A -> B -> C
+    //
+    // Consider the case where B was defined in a different file and C was defined
+    // in the current file: then C will be in our candidate list.
+    //
+    // But if B has a loc in the current file and A was defined in a different file,
+    // then B will also be in our candidate list!
+    //
+    // To avoid that case, we need to walk the ownership chains of each symbol to
+    // deduplicate the candidate list.
+    UnorderedSet<core::SymbolRef> deduplicatedCandidates(candidates.begin(), candidates.end());
+    for (auto ref : candidates) {
+        auto owner = ref.owner(gs);
+        while (owner != core::Symbols::root()) {
+            if (deduplicatedCandidates.contains(owner)) {
+                deduplicatedCandidates.erase(ref);
+                break;
+            }
+
+            owner = owner.owner(gs);
+        }
```

Comment:
It's not clear to me whether I need to be super-defensive and check for `.exists()` or `hideSymbol(owner)` on the way up the chain.